text/0625-helper-managers.md

```diff
@@ -0,0 +1,504 @@
+
+When a value is used as a helper in a template, the helper manager is looked up
+by on the object by walking up its prototype chain and finding the first helper
+manager. This manager then receives the value and can create and manage an
+instance of a helper from it. This provides a layer of indirection that allows
+users to design high-level helper APIs, without Ember needing to worry about the
+details. High-level APIs can be experimented with and iterated on while the
+core of Ember helpers remains stable, and new APIs can be introduced gradually
+over time to existing code bases.
+
+`setHelperManager` receives two arguments:
+
+1. A factory function, which receives the `owner` and returns an instance of a
+   helper manager.
+2. The object to associate the factory function with.
+
+The first time the object is looked up, the factory function will be called to
+create the helper manager. It will be cached, and in subsequent lookups it the
+cached helper manager will be used instead.
+
+Only one helper manager exists per helper factory, so many helpers will end up
+using the same instance of the helper manager. As such, you should not store any
+state on the helper manager that is related to a single helper instance.
+
+Helper managers must fulfill the following interface (This example uses
+[TypeScript interfaces](https://www.typescriptlang.org/docs/handbook/interfaces.html)
+for precision, you do not need to write helper managers using TypeScript):
+
+```ts
+interface HelperManager<HelperStateBucket> {
+  capabilities: HelperCapabilities;
+
+  createHelper(definition: HelperDefinition, args: TemplateArgs): HelperStateBucket;
+
+  getValue?(bucket: HelperStateBucket, args: TemplateArgs): unknown;
+
+  runEffect?(bucket: HelperStateBucket, args: TemplateArgs): void;
+
+  getDestroyable?(bucket: HelperStateBucket): object;
+}
+```
+
+The capabilities property should be provided using the `capabilities()` function
+imported from the same module as `setHelperManager`:
+
+```js
+import { capabilities } from '@ember/helper';
+
+class MyHelperManager {
+  capabilities = capabilities();
+}
+```
+
+Below is a description of each of the methods on the interface and their
+functions.
+
+> The remaining API docs should be copied from the descriptions in the Detailed
+> Design section of this RFC.
+
+#### `capabilities`
+
+`capabilities` returns a capabilities configuration which can be used to modify
+the behavior of the manager. Manager capabilities _must_ be provided using the
+`capabilities` function, as the underlying implementation can change over time.
+
+The first argument to capabilities is a version string, which is the version of
+Ember that the capabilities were defined in. Ember can add new versions at any
+time, and these may have entirely different behaviors, but it will not remove
+old versions until the next major version.
+
+```js
+capabilities('3.x');
+```
+
+The second argument is an optional object of capabilities and boolean values
```

Comment:
I do not actually think this is optional. As defined above, you must either `hasValue` or `isScheduledEffect` to actually do anything.